### ISSUE TITLE
C# client and Server- Possible Memory Leak

### ISSUE THREAD

**Issue by aagarwalincontact** [grpc/grpc#22862]
### What version of gRPC and what language are you using?
C# client and server
Google.Protobuf 3.11.4
Grpc 2.24.0 

### What operating system (Linux, Windows,...) and version?
Windows 10 

### What runtime / compiler are you using (e.g. python version or version of gcc)

.Net Framework 4.7.2

### What did you do?
A server app that implements Service which has one RPC getdata that takes a simple request message that contains a string.
A client app that creates 10 Service clients and then iterates thru the clients and has them call the Getdata RPC with a 2s polling in between each client call. I am reusing the same channel but recreate client for every request. In each call it is returning 1MB data. This is in time polling so basically the clients just continue to call getdata continually ensuring that there is a break of 2 secs inbetween each call to ensure I'm not just backing up the grpc queue.

The memory consumption keep on increasing eventually the server goes out of memory. When profiling the server,I am seeing the live bytes grow and increase is happening in Grpc.Core.Internal. Mostly in RequestCallContextSafeHandle, ReusableSliceBuffer.SliceSegment[], and BatchContextSafeHandle. It says 4 sets with 8000 duplicate instances.

### What did you expect to see?
Server should not go out of memory

### What did you see instead?
Server eventually go out of memory.
Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

**Comment by jtattermusch** [comment]
Can you provide an actual minimal project (ideally as e.g. a git repo) that reproduces the problem? Otherwise it is unlikely I'd be able to reproduce exactly the same setup you used (and without it the issue is not actionable)

**Comment by robgoodwin** [comment]
I am seeing some similar results as the OP.  I will wait a bit to see if a sample is provided, otherwise I can create something.  My application is monitoring service health using the gRPC HealthCheck interface.  I did a quick implementation and just polled, creating and destroying new interfaces every 10 or so seconds to about 10 clients.  I let it run over the weekend and the service started at 30 MB, but at the end of the weekend it was 1.7 GB.  I know that means nothing without a sample, so I will see what I can do, and then test to ensure the sample shows the same issue

**Comment by aagarwalincontact** [comment]
Let me try to write a simple app that replicates the issue and which i can share here. I will revert back in a day or two.

**Comment by aagarwalincontact** [comment]
I create a sample app. Just to simulate production network traffic i m delaying the response randomly between 2, 5 and 10 sec. In client i have kept the deadline of 3 sec. Henceforth, 1/3 of the requests will timeout causing deadline exceeded. When u run the server and client , in 20 - 30 mins app should crash with out of memory exception.
Here is the url for sample app - 
git@github.com:aagarwalincontact/GrpcLoadTest.git

**Comment by yashykt** [comment]
just had a quick look. @jtattermusch does C# provide access to the resource quota? If the user is able to set the resource quota, they'll be able to limit the amount of memory Core will be using

**Comment by jtattermusch** [comment]
@aagarwalincontact  what I see when profiling your app is most of the consumed memory being used by Reply messages (which doesn't inherently seem as something wrong).
The logic of the client seems overly complicated, so I'm not sure what it is trying to achieve.
Basically it seems that each 2seconds, it starts 100 new empty requests, each of which is supposed to generate a response of 10MB (so every 2 seconds, the server would need to send back 2 * 10 * 100 MB of data, so a flow of 1GB/s. it is probably not reasonably to expect that the server would be able send data at that pace - or client to receive it), so I think the memory usage you're seeing might be just from all the concurrent calls you started (each of them allocates a 10MB message).
You are also not throttling the client at all - you're generating the same amount of very cheap empty requests regardless of whether the server can keep up responding to them (which it cannot because the responses are very expensive).

By default the server would allow a large number of concurrent calls (2000 per each completion queue and there can be multiple cqs on multicore machine).
https://github.com/grpc/grpc/blob/44fb37c99f2853cc23f04fba15468980d9e28e41/src/csharp/Grpc.Core/Server.cs#L36
and your app seems to allow each of the handlers to allocate 10MB of memory and doesn't check whether the call was already cancelled (through serverContext.CancellationToken) before generating response of that size.
So basically it seems that your example app is designed in a way that allows several thousand concurrent requests, each one of them allocating 10MB, which I am pretty sure can lead to out of memory error.

Based on what I saw, I don't think there's a problem in the gRPC stack, but in the design of your client/server.

**Comment by jtattermusch** [comment]
Btw, you can check whether the cancellation was already requested by e.g. `context.CancellationToken.IsCancellationRequested` when processing the request on the serverside.
Your server doesn't react adequately to a call being cancelled (basically you generate a 10MB response whether or not the call is already past the deadline or not).

**Comment by jtattermusch** [comment]
Closing as I don't believe there is an issue on the gRPC side.

**Comment by aagarwalincontact** [comment]
Thanks @jtattermusch . This is a sample app where in i deliberately made it to hit 100 concurrent calls and calling in overloaded fashion to simulate network delays and miss. 
As u suggested , we have made the client and server to have throttle limits as well as put the deadline in each request so that if client is not able to receive response due to network, server can expire the response based on deadline. 
Thank u for looking into it .

**Comment by menozz** [comment]
Hi guys, we are faced with the same problem on stress environment:
[](url)
C# client and server
Google.Protobuf 3.14.0
Grpc 2.34.0

Linux (Ubuntu)

.NET Core 5.0
[dump analysis](https://ibb.co/y8HC97Q)

**Comment by Leonardo-Ferreira** [comment]
Having a similar issue here two... bursting my app with requests elevate the memory usage to about 1gb (for example) but removing the load and waiting (for 4-5 hours for example) does not affect the memory at all... `dotnet-dump` cant account for all the memory that is being consumed also

**Comment by Zeleksonob** [comment]
having the same problem here too.

.net 4.8
grpc.core 2.40.0

it seems it's happing when a message contains bytes field send array , lets say of 300 message causes the server process memory to be a least 2gb , when using 1500 message with bytes filed in them the memory jumps up to 20gb and crushes the process
